cookbooks/prescribed_velocity_ascii_data/prescribed_velocity_ascii_data.cc

```diff
@@ -0,0 +1,350 @@
+  Copyright (C) 2011 - 2020 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+#include <deal.II/base/parameter_handler.h>
+#include <deal.II/base/parsed_function.h>
+#include <deal.II/fe/fe_values.h>
+#include <aspect/global.h>
+#include <aspect/utilities.h>
+#include <aspect/simulator_signals.h>
+#include <aspect/initial_composition/ascii_data.h>
+
+namespace aspect
+{
+  using namespace dealii;
+
+  // Global variables (to be set by parameters)
+  bool prescribe_internal_velocities;
+  bool prescribed_velocity_ascii_file_loaded;
+
+  // Because we do not initially know what dimension we're in, we need
+  // function parser objects for both 2d and 3d.
+
+  aspect::Utilities::AsciiDataBase<2> prescribed_velocity_field_params_2d;
+  aspect::Utilities::AsciiDataBase<3> prescribed_velocity_field_params_3d;
+
+  aspect::Utilities::StructuredDataLookup<2> prescribed_velocity_field_data_2d(1, 1.);
+  aspect::Utilities::StructuredDataLookup<3> prescribed_velocity_field_data_3d(1, 1.);
+
+  Functions::ParsedFunction<2> prescribed_velocity_function_2d (2);
+  Functions::ParsedFunction<3> prescribed_velocity_function_3d (3);
+
+  /**
+   * Declare additional parameters.
+   */
+  void declare_parameters(const unsigned int dim,
+                          ParameterHandler &prm)
+  {
+    prm.declare_entry ("Prescribe internal velocities", "false",
+                       Patterns::Bool (),
+                       "Whether or not to use any prescribed internal velocities. "
+                       "Locations in which to prescribe velocities are defined "
+                       "in section ``Prescribed velocities/Indicator function'' "
+                       "and the velocities are defined in section ``Prescribed "
+                       "velocities/Velocity function''. Indicators are evaluated "
+                       "at the center of each cell, and all DOFs associated with "
+                       "the specified velocity component at the indicated cells "
+                       "are constrained."
+                      );
+
+    prm.enter_subsection ("Prescribed velocities");
+    {
+      {
+        if (dim == 2)
+          {
+            prescribed_velocity_field_params_2d.declare_parameters(prm, "blob", "blob", "Ascii data model");
+            //prescribed_velocity_field_data_2d.initialize(1);
```

Comment:
remove the commented out line if you don't need it.

The last 3 arguments of the `declare_parameters` call are funny. Is there a better choice one could pass here?

Separately, the function is (should be) `static`, so maybe `Functions::ParsedFunction<2>::declare_parameters(...);` would be easier to read.

Reply:
This object is only used to read in the location of the data file (and optionally a scale parameter). The middle two arguments are default file directory and filename. I've changed those to be a bit more sensible.